src/Common/SettingsChanges.h

```diff
@@ -8,17 +7,51 @@ namespace DB
 
 class IColumn;
 
-struct SettingChange
+
+struct SettingValue
+{
+    virtual ~SettingValue() = default;
+    virtual const Field & getField() const = 0;
+    virtual Field & getField() = 0;
+    virtual std::string toString() const = 0;
+};
+using SettingValuePtr = std::shared_ptr<SettingValue>;
+
+SettingValuePtr getSettingValueFromField(const Field & field);
+SettingValuePtr getSettingValueFromField(Field && field);
+
+
+class SettingChange
 {
+private:
     String name;
-    Field value;
+    SettingValuePtr value;
```

Comment:
Field is already a variant class. And now we're kind of adding a second-tier variant here. That's not nice.
So I'd prefer if we add a possibility for `Field` to contain an AST like a custom type without knowing what AST is.

Like this

```
class Field
{
   struct ICustomType
   {
       virtual std::unique_ptr<ICustomType> clone() const;
       virtual String toString() const;
       virtual const char * getTypeName() const;
       virtual std::optional<Types::Which> getBaseTypeIfConvertible() const;
       virtual std::optional<Field> tryConvertToBaseType(Types::Which dest_type) const
   };

    struct Types
    {
        /// Type tag.
        enum Which
        {
              ...
              CustomType = 30
        };
```

And then in the clickhouse_parsers target we can inherit `ASTInField` from `Field::ICustomType`.

Reply:
I don't insist that it should be done right now, so if you're in a hurry then just fix conflicts and merge.